portalnet/src/overlay_service.rs

```diff
@@ -2194,26 +2194,53 @@ where
         Ok(closest_nodes)
     }
 
+    /// Returns a vector of ENRs of the `max_nodes` closest connected nodes to the target from our routing table.
+    fn closest_connected_nodes(&self, target_key: &Key<NodeId>, max_nodes: usize) -> Vec<Enr> {
+        // Filter out all disconnected nodes
+        let kbuckets = self.kbuckets.read();
+        let mut all_nodes: Vec<&kbucket::Node<NodeId, Node>> = kbuckets
+            .buckets_iter()
+            .flat_map(|kbucket| {
+                kbucket
+                    .iter()
+                    .filter(|node| node.status.is_connected())
+                    .collect::<Vec<&kbucket::Node<NodeId, Node>>>()
+            })
+            .collect();
+
+        all_nodes.sort_by(|a, b| {
+            let a_distance = a.key.distance(target_key);
+            let b_distance = b.key.distance(target_key);
+            a_distance.cmp(&b_distance)
+        });
+
+        all_nodes
+            .iter()
+            .take(max_nodes)
+            .map(|closest| closest.value.enr.clone())
+            .collect()
+    }
+
     /// Starts a FindNode query to find nodes with IDs closest to `target`.
     fn init_find_nodes_query(
         &mut self,
         target: &NodeId,
         callback: Option<oneshot::Sender<Vec<Enr>>>,
     ) -> Option<QueryId> {
         let target_key = Key::from(*target);
-        let mut closest_enrs: Vec<Enr> = self
-            .kbuckets
-            .write()
-            .closest_values(&target_key)
-            .map(|closest| closest.value.enr)
-            .collect();
 
-        // `closest_enrs` will be empty if `target` is our local node ID
```

Comment:
Do you know if this is still a valid case we must take care of?

Reply:
Yes, we still have to take care of this when all nodes are disconnected, good catch